thermal.c

```diff
@@ -14,43 +14,333 @@
+		    nl_geterror(err->error));
+		*((int *)arg) = err->error;
+	}
+	return NL_SKIP;
+}
+
+static int handle_end(struct nl_msg *msg __attribute__((unused)), void *arg)
+{
+	if (arg)
+		*((int *)arg) = 0;
+	return NL_SKIP;
+}
+
+#define NL_FAMILY_NAME  "nlctrl"
+
+struct msgheader {
+	unsigned char cmd, version;
+	unsigned int port, seq;
+	int id, hdrlen, flags;
+};
 
 static gboolean establish_netlink(void)
 {
-	gboolean error = TRUE;
+	struct msgheader msghdr = { CTRL_CMD_GETFAMILY, 0, 0, 0, 0, 0, 0 };
+	struct family_data nldata = { THERMAL_GENL_EVENT_GROUP_NAME, -ENOENT };
+	struct nl_cb *cloned_callback = NULL;
+	int rc, group_id, callback_rc = 1;
+	struct nl_msg *msg = NULL;
+	gboolean error = FALSE;
+	void *hdr;
+
+	msg = nlmsg_alloc();
+	if (!msg) {
+		log(TO_ALL, LOG_ERR, "thermal: message allocation failed.\n");
+		goto out;
+	}
+
+	msghdr.id = genl_ctrl_resolve(sock, NL_FAMILY_NAME);
+	if (msghdr.id < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: message id enumeration failed.\n");
+		goto err_out;
+	}
+
+	hdr = genlmsg_put(msg, msghdr.port, msghdr.seq, msghdr.id, msghdr.hdrlen,
+			  msghdr.flags, msghdr.cmd, msghdr.version);
+	if (!hdr) {
+		log(TO_ALL, LOG_ERR, "thermal: netlink header setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nla_put_string(msg, CTRL_ATTR_FAMILY_NAME, THERMAL_GENL_FAMILY_NAME);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: message setup failed.\n");
+		goto err_out;
+	}
+
+	cloned_callback = nl_cb_clone(callback);
+	if (!cloned_callback) {
+		log(TO_ALL, LOG_ERR, "thermal: callback handle duplication failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_send_auto_complete(sock, msg);
+	if (rc < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: failed to send the first message.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_err(cloned_callback, NL_CB_CUSTOM, handle_error, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: error callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_ACK, NL_CB_CUSTOM, handle_end, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: ack callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_FINISH, NL_CB_CUSTOM, handle_end, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: finish callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_VALID, NL_CB_CUSTOM, handle_groupid, &nldata);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: group id callback setup failed.\n");
+		goto err_out;
+	}
+
+	while (callback_rc != 0) {
+		rc = nl_recvmsgs(sock, cloned_callback);
+		if (rc < 0) {
+			log(TO_ALL, LOG_ERR, "thermal: failed to receive messages.\n");
+			goto err_out;
+		}
+	}
 
-	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to establish netlink.\n");
+	group_id = nldata.id;
+	if (group_id < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: invalid group_id was received.\n");
+		goto err_out;
+	}
+
+	rc = nl_socket_add_membership(sock, group_id);
```

Comment:
If any of these operations fails, what code is responsible for tearing down the socket?

Reply:
These are in init_thermal() where it falls through deinit_thermal() that frees up sock.